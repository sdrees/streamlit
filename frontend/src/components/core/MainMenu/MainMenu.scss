/**
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@import "src/assets/css/variables";

@keyframes recording-indicator-pulse {
  0% {
    box-shadow: 0 0 0.25rem $red;
  }
  50% {
    box-shadow: 0 0 0.5rem 0.25rem $red;
  }
  100% {
    box-shadow: 0 0 0.25rem $red;
  }
}

#MainMenu {
  .recording-indicator {
    display: block;
    bottom: 1rem;
    right: 0.5rem;
    position: absolute;
    background-color: red;
    width: 0.5rem;
    height: 0.5rem;
    border-radius: 50px;
    box-shadow: 0 0 0.25rem $red;
    animation: recording-indicator-pulse 2s linear infinite;
  }
}

.menu-divider {
  border-top: 1px solid $gray-lightest;
  margin: 0.5rem 0;
}

.menu-item {
  margin: 0;
  padding: 0.25rem 1.5rem;
  display: flex;
  flex-direction: row;
  align-items: flex-start;
  cursor: pointer;

  &.menu-item-stop-recording {
    color: $red;
    font-weight: bolder;

    span.menu-item-shortcut {
      color: $red;
    }
  }

  span.menu-item-shortcut {
    color: $gray;
    font-size: 0.75rem;
    margin-top: 0.25rem;
    font-variant: small-caps;
    text-transform: uppercase;
  }

  span.menu-item-label {
    margin-right: 0.75rem;
    flex-grow: 1;
  }

  &-highlighted {
    background-color: $gray-lightest;
  }

  &-disabled {
    background-color: transparent;
    color: $gray;
    cursor: not-allowed;
  }

  &:active:not(.menu-item-disabled) {
    background-color: $primary;
    color: white;
    outline: none;

    span.menu-item-shortcut {
      color: white;
    }
  }

  &:focus:not(.menu-item-disabled) {
    background-color: $primary;
    color: white;
  }
}
