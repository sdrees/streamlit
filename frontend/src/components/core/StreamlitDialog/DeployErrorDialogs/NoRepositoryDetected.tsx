import React from "react"
import { ONLINE_DOCS_URL, TEAMS_URL } from "urls"
import { IDeployErrorDialog } from "./types"

function NoRepositoryDetected(): IDeployErrorDialog {
  return {
    title: "Unable to deploy app",
    body: (
      <>
        <p>Could not find a remote repository hosted on Github.</p>
        <p>How Streamlit sharing works:</p>
        <ul>
          <li>
            To deploy a public app, you must first put it in a public Github
            repo. See{" "}
            <a
              href={ONLINE_DOCS_URL}
              rel="noopener noreferrer"
              target="_blank"
            >
              our documentation
            </a>{" "}
            for more details.
          </li>
          <li>
            If you'd like to deploy a private app,{" "}
            <a href={TEAMS_URL} target="_blank" rel="noopener noreferrer">
              sign up for Streamlit for Teams
            </a>
            .
          </li>
        </ul>
      </>
    ),
  }
}

export default NoRepositoryDetected
