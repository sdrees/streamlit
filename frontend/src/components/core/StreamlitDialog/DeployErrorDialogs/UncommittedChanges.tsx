import React from "react"
import { IDeployErrorDialog } from "./types"

function UncommittedChanges(module: string): IDeployErrorDialog {
  return {
    title: "Unable to deploy app",
    body: (
      <>
        <p>
          The file <code>{module}</code> has uncommitted changes.
        </p>
        <p>Please commit the latest changes and push to Github to continue.</p>
      </>
    ),
  }
}

export default UncommittedChanges
