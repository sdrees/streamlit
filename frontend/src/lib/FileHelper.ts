import { CancelTokenSource } from "axios"

export interface ExtendedFile extends File {
  id?: string
  status?: string
  errorMessage?: string
  cancelToken?: CancelTokenSource
  progress?: number
}

export enum FileStatuses {
  ERROR = "ERROR",
  DELETING = "DELETING",
  READY = "READY",
  UPLOADING = "UPLOADING",
  UPLOADED = "UPLOADED",
}

export enum FileSizes {
  GigaByte = "gb",
  KiloByte = "kb",
  MegaByte = "mb",
  Byte = "b",
}

export const BYTE_CONVERSION_SIZE = 1024
const sizeUnitSequence = [
  FileSizes.GigaByte,
  FileSizes.MegaByte,
  FileSizes.KiloByte,
  FileSizes.Byte,
]

export const getSizeDisplay = (
  size: number,
  unit: FileSizes,
  rounding = 1
): string => {
  if (!unit) {
    unit = FileSizes.Byte
  }

  if (rounding < 0) {
    rounding = 0
  }

  if (size < 0) {
    throw new Error("Size must be greater than or equal to 0")
  }

  const sizeIndex = sizeUnitSequence.indexOf(unit)
  const nextUnitSize = size / BYTE_CONVERSION_SIZE
  if (sizeIndex && size > BYTE_CONVERSION_SIZE / 2) {
    return getSizeDisplay(
      nextUnitSize,
      sizeUnitSequence[sizeIndex - 1],
      rounding
    )
  }
  return `${size.toFixed(rounding)}${unit.toUpperCase()}`
}

export const sizeConverter = (
  size: number,
  inputUnit: FileSizes,
  outputUnit: FileSizes
): number => {
  if (size < 0) {
    throw Error("Size must be 0 or greater")
  }

  const inputLevel = sizeUnitSequence.findIndex(unit => unit === inputUnit)
  const outputLevel = sizeUnitSequence.findIndex(unit => unit === outputUnit)

  if (inputLevel === -1 || outputLevel === -1) {
    // Should not ever occur
    throw Error("Unexpected byte unit provided")
  }

  if (inputLevel === outputLevel) {
    return size
  }

  const levelsBetween = Math.abs(inputLevel - outputLevel)
  const byteDifference = BYTE_CONVERSION_SIZE ** levelsBetween

  if (inputLevel > outputLevel) {
    // Going from smaller to bigger
    return size / byteDifference
  }
  // Going from bigger to smaller
  return size * byteDifference
}
