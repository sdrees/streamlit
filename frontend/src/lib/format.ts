/**
 * @license
 * Copyright 2018-2021 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Text formatting utilities
 */
import moment from "moment"
import "moment-duration-format"
import numbro from "numbro"

class Duration {
  private readonly millis: number

  constructor(millis: number) {
    this.millis = millis
  }

  getTime(): number {
    return this.millis
  }
}

class Format {
  static nanosToDate(nanos: number): Date {
    return new Date(nanos / 1e6)
  }

  static iso8601ContainsTimezone(iso: string): boolean {
    // Moment doesn't provide a method to determine if an iso string has a
    // timezone. By the time the string is parsed into a moment object, the
    // timezone will have already been set to UTC or localtime.
    //
    // What we can do is call moment.parseZone() and moment.utc() on the string.
    // If the string has a timezone, the resulting datetimes will be different.
    // If it doesn't, the resulting datetimes will be the same.
    const a = moment.parseZone(iso)
    const b = moment.utc(iso)

    // Passing `true` to the `.utc()` method of a moment object normalizes its
    // timezone to UTC changing the datetime. "1/1/1970 midnight at PST-8"
    // becomes "1/1/1970 midnight at UTC".  Since we want to know if the two
    // datetimes are the same, normalize both to UTC, then compare.
    return !a.utc(true).isSame(b.utc(true))
  }

  static iso8601ToMoment(iso: string): moment.Moment {
    return Format.iso8601ContainsTimezone(iso)
      ? moment.parseZone(iso)
      : moment(iso)
  }

  static nanosToDuration(nanos: number): Duration {
    return new Duration(nanos / 1e6)
  }

  static dateToString(date: Date): string {
    const m = moment(date)
    let format = "lll"
    if (m.hour() === 0 && m.minute() === 0 && m.second() === 0) {
      format = "ll"
    }
    return m.format(format)
  }

  static durationToString(duration: Duration): string {
    const ms = moment.duration(duration.getTime()).asMilliseconds()
    return moment.utc(ms).format()
  }
}

/**
 * Formats the string nicely if it's a floating point, number, date or duration.
 */
function toFormattedString(x: any): string {
  if (moment.isMoment(x)) {
    return x.format()
  }
  if (isFloat(x)) {
    return numbro(x).format("0,0.0000")
  }
  if (x instanceof Date) {
    return Format.dateToString(x)
  }
  if (x instanceof Duration) {
    return Format.durationToString(x)
  }
  return x.toString()
}

/**
 * Returns true if this number is a float.
 */
function isFloat(n: any): boolean {
  return Number(n) === n && n % 1 !== 0
}

export { Duration, Format, toFormattedString }
