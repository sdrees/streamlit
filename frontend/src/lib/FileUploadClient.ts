/**
 * @license
 * Copyright 2018-2021 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CancelToken } from "axios"
import { ExtendedFile } from "lib/FileHelper"
import HttpClient from "lib/HttpClient"
import { SessionInfo } from "lib/SessionInfo"

/**
 * Handles uploading files to the server.
 */
export class FileUploadClient extends HttpClient {
  /**
   * Upload a file to the server. It will be associated with this browser's sessionID.
   *
   * @param widgetId: the ID of the FileUploader widget that's doing the upload.
   * @param files: the files to upload.
   * @param onUploadProgress: an optional function that will be called repeatedly with progress events during the upload.
   * @param cancelToken: an optional axios CancelToken that can be used to cancel the in-progress upload.
   * @param replace: an optional boolean to indicate if the file should replace existing files associated with the widget.
   */
  public async uploadFiles(
    widgetId: string,
    files: ExtendedFile[],
    totalFiles?: number,
    onUploadProgress?: (progressEvent: any) => void,
    cancelToken?: CancelToken,
    replace?: boolean
  ): Promise<void> {
    const form = new FormData()
    form.append("sessionId", SessionInfo.current.sessionId)
    form.append("widgetId", widgetId)

    // We need to send totalFiles in order to reduce reruns for multiple file uploads.
    // We are uploading files in parallel so that if one file fails, the rest do not.
    // Because these are happening in parallel, the server needs to know how many files
    // are expected before trigger a rerun as reruns can be expensive.
    form.append("totalFiles", (totalFiles || files.length).toString())
    if (replace) form.append("replace", "true")
    for (const file of files) {
      form.append(file.id || file.name, file, file.name)
    }

    await this.request("upload_file", {
      cancelToken,
      method: "POST",
      data: form,
      onUploadProgress,
    })
  }

  public async delete(widgetId: string, fileId: string): Promise<void> {
    await this.request(
      `upload_file/${SessionInfo.current.sessionId}/${widgetId}/${fileId}`,
      { method: "DELETE" }
    )
  }
}
