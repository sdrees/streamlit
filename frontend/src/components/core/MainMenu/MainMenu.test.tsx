/**
 * @license
 * Copyright 2018-2020 Streamlit Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react"
import { shallow } from "lib/test_util"
import { IMenuItem } from "hocs/withS4ACommunication/types"
import { Args as SessionInfoArgs, SessionInfo } from "lib/SessionInfo"

import { GitInfo, IGitInfo } from "autogen/proto"
import { IDeployErrorDialog } from "components/core/StreamlitDialog/DeployErrorDialogs/types"
import {
  DetachedHead,
  ModuleIsNotAdded,
  NoRepositoryDetected,
  RepoIsAhead,
  UncommittedChanges,
  UntrackedFiles,
} from "components/core/StreamlitDialog/DeployErrorDialogs"

import MainMenu, { Props } from "./MainMenu"

const { GitStates } = GitInfo

const getProps = (extend?: Partial<Props>): Props => ({
  aboutCallback: jest.fn(),
  clearCacheCallback: jest.fn(),
  isServerConnected: true,
  quickRerunCallback: jest.fn(),
  s4aMenuItems: [],
  screencastCallback: jest.fn(),
  screenCastState: "",
  sendS4AMessage: jest.fn(),
  settingsCallback: jest.fn(),
  shareCallback: jest.fn(),
  sharingEnabled: false,
  isDeployErrorModalOpen: false,
  showDeployError: jest.fn(),
  loadGitInfo: jest.fn(),
  closeDialog: jest.fn(),
  ...extend,
})

describe("App", () => {
  beforeAll(() => {
    SessionInfo.current = new SessionInfo({
      sessionId: "sessionId",
      streamlitVersion: "sv",
      pythonVersion: "pv",
      installationId: "iid",
      installationIdV1: "iid1",
      installationIdV2: "iid2",
      authorEmail: "ae",
      maxCachedMessageAge: 2,
      commandLine: "command line",
      userMapboxToken: "mpt",
    } as SessionInfoArgs)
  })

  it("renders without crashing", () => {
    const props = getProps()
    const wrapper = shallow(<MainMenu {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it("should render s4a menu items", () => {
    const items: IMenuItem[] = [
      {
        type: "text",
        label: "Share this app",
        key: "share",
      },
      {
        type: "separator",
      },
      {
        type: "text",
        label: "View app source",
        key: "source",
      },
      {
        type: "text",
        label: "Report bug with app",
        key: "support",
      },
      {
        type: "separator",
      },
      {
        type: "text",
        label: "About Streamlit for Teams",
        key: "about",
      },
    ]
    const props = getProps({
      s4aMenuItems: items,
    })
    const wrapper = shallow(<MainMenu {...props} />)
    const popoverContent = wrapper.find("StatefulPopover").prop("content")

    // @ts-ignore
    const menuWrapper = shallow(popoverContent(() => {})).dive()

    // @ts-ignore
    const menuLabels = menuWrapper
      .find("MenuStatefulContainer")
      .prop("items")
      // @ts-ignore
      .map(item => item.label)
    expect(menuLabels).toEqual([
      "Rerun",
      "Settings",
      "Record a screencast",
      "Share this app",
      "View app source",
      "Report bug with app",
      "About Streamlit for Teams",
    ])
  })

  it("should render core set of menu elements", () => {
    const props = getProps()
    const wrapper = shallow(<MainMenu {...props} />)
    const popoverContent = wrapper.find("StatefulPopover").prop("content")
    // @ts-ignore
    const menuWrapper = shallow(popoverContent(() => {})).dive()

    // @ts-ignore
    const menuLabels = menuWrapper
      .find("MenuStatefulContainer")
      .prop("items")
      // @ts-ignore
      .map(item => item.label)
    expect(menuLabels).toEqual([
      "Rerun",
      "Clear cache",
      "Deploy this app",
      "Record a screencast",
      "Documentation",
      "Ask a question",
      "Report a bug",
      "Streamlit for Teams",
      "Settings",
      "About",
    ])
  })

  it("should render deploy app menu item", () => {
    const props = getProps({ gitInfo: {} })
    const wrapper = shallow(<MainMenu {...props} />)
    const popoverContent = wrapper.find("StatefulPopover").prop("content")
    // @ts-ignore
    const menuWrapper = shallow(popoverContent(() => {})).dive()

    // @ts-ignore
    const menuLabels = menuWrapper
      .find("MenuStatefulContainer")
      .prop("items")
      // @ts-ignore
      .map(item => item.label)
    expect(menuLabels).toEqual([
      "Rerun",
      "Clear cache",
      "Deploy this app",
      "Record a screencast",
      "Documentation",
      "Ask a question",
      "Report a bug",
      "Streamlit for Teams",
      "Settings",
      "About",
    ])
  })

  describe("Onclick deploy button", () => {
    function testDeployErrorModal(
      gitInfo: Partial<IGitInfo>,
      dialogComponent: (module: string) => IDeployErrorDialog
    ): void {
      const props = getProps({
        gitInfo,
      })
      const wrapper = shallow(<MainMenu {...props} />)
      const popoverContent = wrapper.find("StatefulPopover").prop("content")
      // @ts-ignore
      const menuWrapper = shallow(popoverContent(() => {})).dive()

      const items: any = menuWrapper.prop("items")

      const deployOption = items.find(
        // @ts-ignore
        ({ label }) => label === "Deploy this app"
      )

      deployOption.onClick()

      // @ts-ignore
      const dialog = dialogComponent(props.gitInfo.module)

      expect(props.showDeployError.mock.calls[0][0]).toStrictEqual(
        dialog.title
      )
      expect(props.showDeployError.mock.calls[0][1]).toStrictEqual(dialog.body)
    }

    it("no repo or remote", () => {
      testDeployErrorModal(
        {
          state: GitStates.DEFAULT,
        },
        NoRepositoryDetected
      )
    })

    it("empty repo", () => {
      testDeployErrorModal(
        {
          repository: "",
          branch: "",
          module: "",
          state: GitStates.DEFAULT,
        },
        NoRepositoryDetected
      )
    })

    it("repo is detached", () => {
      testDeployErrorModal(
        {
          repository: "repo",
          branch: "branch",
          module: "module",
          state: GitStates.HEAD_DETACHED,
        },
        DetachedHead
      )
    })

    it("script was not added to the repo", () => {
      testDeployErrorModal(
        {
          repository: "repo",
          branch: "branch",
          module: "module.py",
          isHeadDetached: false,
          untrackedFiles: ["module.py"],
        },
        ModuleIsNotAdded
      )
    })

    it("uncommitted changes", () => {
      testDeployErrorModal(
        {
          repository: "repo",
          branch: "branch",
          module: "module.py",
          isHeadDetached: false,
          uncommittedFiles: ["module.py"],
          untrackedFiles: [],
        },
        UncommittedChanges
      )
    })

    it("changes not pushed to Github", () => {
      const deployParams: IGitInfo = {
        repository: "repo",
        branch: "branch",
        module: "module.py",
        uncommittedFiles: [],
        untrackedFiles: [],
        state: GitStates.AHEAD_OF_REMOTE,
      }
      testDeployErrorModal(deployParams, RepoIsAhead)
    })

    it("untracked files", () => {
      testDeployErrorModal(
        {
          repository: "repo",
          branch: "branch",
          module: "module.py",
          isHeadDetached: false,
          untrackedFiles: ["another-file.py"],
        },
        UntrackedFiles
      )
    })
  })
})
