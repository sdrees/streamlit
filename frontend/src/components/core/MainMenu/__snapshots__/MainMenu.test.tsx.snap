// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
<StatefulPopover
  accessibilityType="menu"
  content={[Function]}
  dismissOnClickOutside={true}
  dismissOnEsc={true}
  focusLock={true}
  ignoreBoundary={false}
  onMouseEnterDelay={200}
  onMouseLeaveDelay={200}
  overrides={
    Object {
      "Body": Object {
        "props": Object {
          "data-test": "main-menu-popover",
        },
      },
    }
  }
  placement="bottomRight"
  showArrow={false}
  stateReducer={[Function]}
  triggerType="click"
>
  <span
    id="MainMenu"
  >
    <Button
      disabled={false}
      fluidWidth={false}
      kind="icon"
      size="medium"
    >
      <Icon
        type="menu"
      />
    </Button>
  </span>
</StatefulPopover>
`;
